### PR TITLE
tsnet: add Tailscale-as-a-library package

### PR DESCRIPTION
**Pull request by bradfitz**

### CODE REVIEW

File: tsnet/tsnet.go [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+		return nil, false
+	}
+	return &apitype.WhoIsResponse{
+		Node:        n,
+		UserProfile: &up,
+	}, true
+}
+
+func (s *Server) doInit() {
+	s.initErr = s.start()
```
Comment (crawshaw) [inline tsnet/tsnet.go:86]: if s.initErr != nil { s.initErr = fmt.Errorf("tsnet: %w", s.initErr) }

File: tsnet/tsnet.go [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+	if err := ns.Start(); err != nil {
+		return fmt.Errorf("failed to start netstack: %w", err)
+	}
+
+	statePath := filepath.Join(s.dir, "tailscaled.state")
+	store, err := ipn.NewFileStore(statePath)
+	if err != nil {
+		return err
+	}
+	logid := "tslib-TODO"
```
Comment (crawshaw) [inline tsnet/tsnet.go:162]: Interesting problem. Just generate a new one every time right now?

File: tsnet/tsnet.go [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+	case ln.conn <- c:
+	case <-t.C:
+		c.Close()
+	}
+}
+
+func (s *Server) Listen(network, addr string) (net.Listener, error) {
+	host, port, err := net.SplitHostPort(addr)
+	if err != nil {
+		return nil, err
```
Comment (crawshaw) [inline tsnet/tsnet.go:208]: fmt.Errorf("tsnet: %w", err)

File: tsnet/tsnet.go [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+
+		conn: make(chan net.Conn),
+	}
+	s.mu.Lock()
+	if s.listeners == nil {
+		s.listeners = map[listenKey]*listener{}
+	}
+	if _, ok := s.listeners[key]; ok {
+		s.mu.Unlock()
+		return nil, fmt.Errorf("listener already open for %s, %s", network, addr)
```
Comment (crawshaw) [inline tsnet/tsnet.go:230]: tsnet:

File: tsnet/tsnet.go [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+	s    *Server
+	key  listenKey
+	addr string
+	conn chan net.Conn
+}
+
+func (ln *listener) Accept() (net.Conn, error) {
+	c, ok := <-ln.conn
+	if !ok {
+		return nil, net.ErrClosed
```
Comment (crawshaw) [inline tsnet/tsnet.go:253]: fmt.Errorf("tsnet: %w", net.ErrClosed)

**Review by crawshaw — APPROVED** [review]

File: control/controlclient/direct.go [hunk 6]
```diff
@@ -177,7 +177,10 @@ func NewDirect(opts Options) (*Direct, error) {
 var osVersion func() string // non-nil on some platforms
 
 func NewHostinfo() *tailcfg.Hostinfo {
-	hostname, _ := os.Hostname()
+	hostname := os.Getenv("TAILSCALE_HOSTNAME")
```
[the remark this answers is not in the document]
Comment (bradfitz) [reply control/controlclient/direct.go]: removed. Remembered the better way to do it.

File: tsnet/tsnet.go (hunk 2 above)
Comment (bradfitz) [reply tsnet/tsnet.go:162]: That's not the hard part. See the TODO:

```
+       // TODO(bradfitz): start logtail? don't use filch, perhaps?
+       // only upload plumbed Logf?
```

We don't want to capture their application's own log.Printf logs.

**Pull request merged**